pkg/pii/pii.go

```diff
@@ -0,0 +1,136 @@
+	"%DOWNSTREAM_REMOTE_ADDRESS_WITHOUT_PORT%": true,
+	"%DOWNSTREAM_LOCAL_ADDRESS%":               true,
+	"%DOWNSTREAM_LOCAL_ADDRESS_WITHOUT_PORT%":  true,
+	"%REQUESTED_SERVER_NAME%":                  true,
+	"%DOWNSTREAM_LOCAL_URI_SAN%":               true,
+	"%DOWNSTREAM_PEER_URI_SAN%":                true,
+	"%DOWNSTREAM_LOCAL_SUBJECT%":               true,
+	"%DOWNSTREAM_PEER_SUBJECT%":                true,
+	"%DOWNSTREAM_PEER_ISSUER%":                 true,
+	"%DOWNSTREAM_TLS_SESSION_ID%":              true,
+	"%DOWNSTREAM_TLS_CIPHER%":                  true,
+	"%DOWNSTREAM_TLS_VERSION%":                 true,
+	"%DOWNSTREAM_PEER_FINGERPRINT_256%":        true,
+	"%DOWNSTREAM_PEER_SERIAL%":                 true,
+	"%DOWNSTREAM_PEER_CERT%":                   true,
+	"%DOWNSTREAM_PEER_CERT_V_START%":           true,
+	"%DOWNSTREAM_PEER_CERT_V_END%":             true,
+	// keys with variables
+	"%REQ(X?Y):Z%":                         true,
+	"%RESP(X?Y):Z%":                        true,
+	"%TRAILER(X?Y):Z%":                     true,
+	"%DYNAMIC_METADATA(NAMESPACE:KEY*):Z%": true,
+}
+
+// Filter filters log fields using pii and modify all PII fields using f
+type Filter struct {
+	f      func(string) string
+	pii    map[string]bool
+	format []string
+}
+
+// NewFilter constructs a custom filter object
+func NewFilter(formatStr string, hash func(string) string, pii map[string]bool) (Filter, error) {
+	// splitting formats and handle error
+	format, ok := shell.Split(formatStr)
+	if !ok {
+		return Filter{}, fmt.Errorf("error in splitting format string: %s", format)
+	}
+	return Filter{f: hash, pii: pii, format: format}, nil
+}
+
+// Default creates a filter with default fields for istio
+func Default() (Filter, error) {
+	return NewFilter(istioFormat, defaultHash, defaultPII)
+}
+
+// Process logs and hash the filter, assumes that filter has valid fields
+func (f Filter) Process(logs []string) []string {
+	out := make([]string, 0, len(logs))
+	for _, log := range logs {
+		fieldValues, ok := shell.Split(log)
+		if !ok {
+			if logger.DebugEnabled() {
+				logger.Debugf("error splitting log, skipping: %s", log)
+			} else {
+				logger.Info("error splitting log, skipping")
+			}
+			continue
+		}
+
+		if len(fieldValues) == len(f.format) {
+			// pick the PII fields and Hash the fields
+			for j, name := range f.format {
+				// TODO: may require further parsing name to verify if it matches as a key in the map with variables
+				if f.pii[name] {
+					fieldValues[j] = f.f(fieldValues[j])
+				}
+			}
+			out = append(out, shell.Join(fieldValues))
+		}
+	}
+	return out
+}
+
+// ProcessLogs process logs with the default filter
+// an empty array of logs and an error instance is returned in the event of an error
+func ProcessLogs(logs []string) ([]string, error) {
+	return defaultFilter.Process(logs), nil
+}
+
+// defaultHash returns the hashed value of s using sha256 defaultHash function
+// TODO: salt the Hash
+//nolint: unused,deadcode
```

Comment:
This should not be needed, right? (general rule is that we want to avoid directives in source code as much as possible as they rapidly go out of date)